source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+
+.. code-block:: none
+
+   // Shift phase of qubit 0's frame by pi/4, eg. an rz gate with angle -pi/4
+   shift_phase transmit(%0), pi/4;
+
+   // Equivalent, but more verbose
+   frame fr = frameof(transmit(%0));
+   shift_phase fr, pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
+     shift_phase transmit(%q), -theta;
+   }
+
+Here's an example qubit spectroscopy experiment.
+
+.. code-block:: none
+
+   qubit q;
+
+   const shots = 1000;
+   const dfreq = 1e6; # 1MHz per point
+   const points = 50; # Sweep over 50MHz
+
+   complex[32] iq, average;
+   complex[32] output[points];
+
+   for p in [0 : points-1] {
+     average = 0;
+     for i in [0 : shots-1] {
+       // Assumes suitable calibrations for reset, x, and measure_iq
+       reset q;
+       x q;
+       measure_iq q -> iq;
+
+       average = (average * i + iq) / (i + 1);
+     }
+     shift q;
+     output[p] = average;
+   }
+
+   defcal prelude {
+      set_frequency transmit(%q) 5e9;
+   }
+
+   defcal shift %q {
+     shift_frequency transmit(%q) dfreq;
+   }
+
+Pulses
+------
+
+Pulses have two representations:
+
+- An array of complex samples which define the points for the pulse envelope
+- A``pulse`` type, which describes an abstract mathematical function
+  representing a pulse. This will later be materialized into a list of complex
+  samples, either by the compiler or the hardware using the parameters provided
+  to the pulse template.
+
+A value of type ``pulse`` is retrieved by calling one of the built-in pulse
+template functions. Note that each of these functions takes a type ``length``
+as a first argument, since pulses need to have a definite length. Using the
+hardware dependent ``dt`` unit is recommended, since the compiler may need to
+down-sample a higher precision pulse to physically realize it.
+
+.. code-block:: none
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   gaussian(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   sech(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // square_width is the width of the square pulse component
+   // sigma is the standard deviation of pulse
+   gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   // beta is the Y correction amplitude, see the DRAG paper
+   drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
+
+   // Define a boxcar (aka. constant) pulse of length l
+   boxcar(l:length)
```

Comment:
I would argue `constant` may be more immediately recognizable but this may be my own bias.